### PR TITLE
Emit kythe/tag/deprecated for elements annotated @Deprecated

### PR DESCRIPTION
**Pull request by cushon**
Follow-up to 84a77bf9671833b3044c390f5b565fa7464aefff.

### CODE REVIEW

**Comment by cushon** [comment]
The way I'm passing the data between `KytheDocTreeScanner` and `KytheTreeScanner` isn't very clean; feedback on a preferred way to structure that would be welcome.

**Comment by cushon** [comment]
> What about KytheDocTreeScanner::visitDocComment returning a richer object than a boolean, nominally a autovalue-like DocCommentVisitResult that wraps the boolean and the deprecation message?

Done, thanks

File: kythe/java/com/google/devtools/kythe/analyzers/java/KytheDocTreeScanner.java [hunk 1]
```diff
@@ -56,11 +57,23 @@ public KytheDocTreeScanner(KytheTreeScanner treeScanner, Context context) {
     this.trees = JavacTrees.instance(context);
   }
 
-  public boolean visitDocComment(TreePath treePath, VName node, EntrySet absNode) {
+  @AutoValue
+  abstract static class DocCommentVisitResult {
+    abstract boolean documented();
+
+    abstract Optional<String> deprecated();
```
Comment (jaysachs) [inline kythe/java/com/google/devtools/kythe/analyzers/java/KytheDocTreeScanner.java]: deprecation() ? Seems a bit odd to have a non-boolean named "deprecated".

File: kythe/java/com/google/devtools/kythe/analyzers/java/KytheDocTreeScanner.java [hunk 2]
```diff
@@ -86,10 +99,9 @@ public int transform(int pos) {
         JAVADOC,
         bracketed,
         anchoredTo,
-        deprecated,
         node,
         absNode == null ? null : absNode.getVName());
-    return true;
+    return DocCommentVisitResult.create(/* documented= */ true, /* deprecated= */ deprecated);
```
Comment (jaysachs) [inline kythe/java/com/google/devtools/kythe/analyzers/java/KytheDocTreeScanner.java]: Can the "deprecated" field just be a local now?

File: kythe/javatests/com/google/devtools/kythe/analyzers/java/testdata/pkg/Deprecation.java [hunk 3]
```diff
@@ -1,25 +1,28 @@
...
-//- @+8Deprecation defines/binding DeprecationClass
+//- @+9Deprecation defines/binding DeprecationClass
 
 /**
  * This the javadoc summary fragment, not the {@code @deprecated} tag.
  *
  * @deprecated this class is obsolete; prefer {@link Comments}
  * @author this is an author tag, not the {@code @deprecated} tag.
  */
+@Deprecated
```
Comment (jaysachs) [inline kythe/javatests/com/google/devtools/kythe/analyzers/java/testdata/pkg/Deprecation.java]: I thought this file was to test deprecated javadoc in the absence of the annotation?

File: kythe/java/com/google/devtools/kythe/analyzers/java/KytheDocTreeScanner.java (hunk 2 above)
Comment (cushon) [reply kythe/java/com/google/devtools/kythe/analyzers/java/KytheDocTreeScanner.java]: Not without additional refactoring, the field is set in `KytheDocTreeScanner#visitDeprecated`.

File: kythe/java/com/google/devtools/kythe/analyzers/java/KytheDocTreeScanner.java (hunk 1 above)
Comment (cushon) [reply kythe/java/com/google/devtools/kythe/analyzers/java/KytheDocTreeScanner.java]: Sure, done

File: kythe/javatests/com/google/devtools/kythe/analyzers/java/testdata/pkg/Deprecation.java (hunk 3 above)
Comment (cushon) [reply kythe/javatests/com/google/devtools/kythe/analyzers/java/testdata/pkg/Deprecation.java]: I wanted to confirm that we save the text from the javadoc if both the javadoc and the annotation are present.

The `deprecatedFieldEmpty` example below still exercises the case where there's javadoc but no annotation.

So I think this has OK coverage of the current logic, but I can split this out into more test cases if you'd prefer?

File: kythe/java/com/google/devtools/kythe/analyzers/java/KytheDocTreeScanner.java [hunk 4]
```diff
@@ -56,15 +57,27 @@ public KytheDocTreeScanner(KytheTreeScanner treeScanner, Context context) {
...
   }
 
-  public boolean visitDocComment(TreePath treePath, VName node, EntrySet absNode) {
+  @AutoValue
+  abstract static class DocCommentVisitResult {
+    abstract boolean documented();
+
+    abstract Optional<String> deprecation();
+
+    static DocCommentVisitResult create(boolean documented, Optional<String> deprecation) {
```
Comment (jaysachs) [inline kythe/java/com/google/devtools/kythe/analyzers/java/KytheDocTreeScanner.java]: Let's optimize a bit. Have a static constant UNDOCUMENT, and one creation method that takes the Optional<String>.

File: kythe/javatests/com/google/devtools/kythe/analyzers/java/testdata/pkg/Deprecation.java (hunk 3 above)
Comment (jaysachs) [reply kythe/javatests/com/google/devtools/kythe/analyzers/java/testdata/pkg/Deprecation.java]: Yes to preserving the javadoc deprecation text if both are present.

Test coverage LGTM.

File: kythe/java/com/google/devtools/kythe/analyzers/java/KytheDocTreeScanner.java (hunk 4 above)
Comment (cushon) [reply kythe/java/com/google/devtools/kythe/analyzers/java/KytheDocTreeScanner.java]: Done

**Pull request merged**